File: course_android/LaunchMode_SignleInstance/app/src/main/java/com/better/learn/launchmode1/OtherActivity.java
package com.better.learn.launchmode1;

import android.os.Bundle;

import androidx.appcompat.app.AppCompatActivity;

public class OtherActivity extends AppCompatActivity {

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_other);
        // singleInstance 启动standard Activity 并不会直接放入到singleInstance任务栈中，因为singleInstance是只能放一个Activity。
        // 所以这个Activity 放入是 包名所在的任务栈，点击返回就直接看到主页的。
    }
}

File: course_android/SharedMemory/app/src/main/java/com/better/learn/sharedmem/OtherActivity.kt
package com.better.learn.sharedmem

import android.graphics.BitmapFactory
import android.os.Build
import androidx.appcompat.app.AppCompatActivity
import android.os.Bundle
import android.os.Messenger
import android.os.SharedMemory
import android.util.Log
import android.widget.ImageView
import androidx.annotation.RequiresApi

class OtherActivity : AppCompatActivity() {
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_other)

        if (Build.VERSION.SDK_INT >= Build.VERSION_CODES.O_MR1) {
            val binderProxy = intent.extras?.getBinder("hello")
            if (null == binderProxy) {
                Log.d("Better", "shit null binder")
                return
            }
            val binder = IMyAidlInterface.Stub.asInterface(binderProxy)
            handleData(binder.get())
        }
    }

    @RequiresApi(Build.VERSION_CODES.O_MR1)
    private fun handleData(sm: SharedMemory) {
        val imageView = findViewById<ImageView>(R.id.imageview)
        val byteBuffer = sm.mapReadOnly()
        // 获取数据
        val len = byteBuffer.limit() - byteBuffer.position()
        val byteArray = ByteArray(len)
        byteBuffer.get(byteArray)
        // 转换
        val bitmap = BitmapFactory.decodeByteArray(byteArray, 0, len)
        imageView.setImageBitmap(bitmap)
        // 释放
        SharedMemory.unmap(byteBuffer)
        sm.close()
    }
}